File: UniTap/app/src/main/java/com/unitap/unitap/NFCBackend/Packetization/CRC16.java
package com.unitap.unitap.NFCBackend.Packetization;

/**
 * A Full CRC Generator
 * Created by Brandon Marino on 2015-11-26.
 */
public class CRC16 {
    /**
     * Generate a CRC from a byte array
     * @param input the byte array to generate the CRC on
     * @return the int of the CRC (CRC is in the two lowest bytes)
     */
    private static int generateCrc(final byte[] input) {
        /*Format: CRC-CCITT (0xFFFF)*/
        int crc = 0xFFFF;

        for (int j = 0; j < input.length ; j++) {
            crc = ((crc  >>> 8) | (crc  << 8) )& 0xffff;
            crc ^= (input[j] & 0xff);//byte to int, trunc sign
            crc ^= ((crc & 0xff) >> 4);
            crc ^= (crc << 12) & 0xffff;
            crc ^= ((crc & 0xFF) << 5) & 0xffff;
        }
        crc &= 0xffff;
        return crc;
    }

    /**
     * This function will take a full byte array, compare it's original Two bytes against the rest of the message
     * The entirety of the rest of the packet will be error checked. Nothing should have been changed
     * @param input int byte array with the appended crc
     * @return if the crc in the first couple of bytes
     */
    protected static boolean verify(byte[] input){
        //now cut it down to remove the crc
        byte[] crcBytesOld = {input[0], input[1]};
        //now combine into an integer for easy comparison.
        int crcIntegerOld = combineBytes(crcBytesOld);
        //now cut the actual message out of the input and generate CRC from that message
        byte[] originalMessage = new byte[input.length-2];
        System.arraycopy(input, 2, originalMessage, 0, input.length-2);
        int crcIntegerNew = generateCrc(originalMessage);
        return crcIntegerOld == crcIntegerNew;
    }

    /**
     * Combine two bytes into one integer
     * @param bytes the two bytes that should be in the least significant bytes of the integer
     * @return the combined int
     */
    protected static int combineBytes(byte [] bytes){
        if (bytes.length <= 8){
            return bytes[1]<<8 &0xFF00 | bytes[0]&0xFF;
        }
        return -1;
    }

    /**
     * Take a message and add a crc of the message to the beginning
     * @param input the original message
     * @return the message of the appended crc
     */
    protected static byte[] appendCRCBytes(byte[] input){
        int crcInt = generateCrc(input); // check

        //mask out the two important bytes
        byte[] crcBytes = new byte[2];
        crcBytes[0] = (byte) (crcInt & 0x000000ff);
        crcBytes[1] = (byte) ((crcInt & 0x0000ff00) >>> 8);

        //now merge those bytes with the original message
        byte [] output = new byte[crcBytes.length+input.length];
        System.arraycopy(crcBytes, 0, output, 0, crcBytes.length);
        System.arraycopy(input, 0, output, crcBytes.length, input.length);

        //return the concatenated message
        return output;
    }

    /**
     * Take a message that has a CRC attached to its head and extract out the message behind it
     * @param appendedMessage take out the crc header ang get the rest of the message
     * @return the original message without the crc
     */
    protected static String stripOutCRCHeader(byte[] appendedMessage){

        byte [] receivedMessage = new byte[appendedMessage.length-2];
        System.arraycopy(appendedMessage, 2, receivedMessage, 0, appendedMessage.length-2);
        return new String(receivedMessage);
    }

    /**
     * This will take one message (in byte array 1), and append a CRC generated from another (int byte array 2)
     * @param arrayToAppend the byte array to add the CRC to
     * @param arrayToCRC the byte array to generate the CRC from
     * @return the appended array
     */
    protected static byte[] appendCRCOnDifferentArray(byte[] arrayToAppend, byte[] arrayToCRC){
        int crcInt = generateCrc(arrayToCRC); // check

        //mask out the two important bytes
        byte[] crcBytes = new byte[2];
        crcBytes[0] = (byte) (crcInt & 0x000000ff);
        crcBytes[1] = (byte) ((crcInt & 0x0000ff00) >>> 8);

        //now merge those bytes with the original message
        byte [] output = new byte[crcBytes.length+arrayToAppend.length];
        System.arraycopy(crcBytes, 0, output, 0, crcBytes.length);
        System.arraycopy(arrayToAppend, 0, output, crcBytes.length, arrayToAppend.length);

        //return the concatenated message
        return output;
    }
}

File: UniTap Terminal/HCEandTag/HCE-BackEnd/CRC/CRC16.c
#include <stdint.h>
#define HCE_MAX_MESSAGE 31
#define OVERHEAD 9
int combineBytes(uint8_t input[]);

/**
 * A Full CRC Generator
 * Created by Brandon Marino on 2015-11-26.
 */

int generateUniTapCrc(uint8_t input[],int sizeOfArray)
{
	uint16_t crc = 0xffff;
	for(int i = 0; i < sizeOfArray; i++){
   		crc = crc ^ (input[i] << 8);
 		for (int j=0; j<8; j++){
   			if (crc & 0x8000)
     			crc = (crc << 1) ^ 0x1021;
   			else
     			crc <<= 1;
     	}
 	}
	return crc;
}

    /**
     * This function will take a full byte array, compare it's original Two bytes against the rest of the message
     * The entirety of the rest of the packet will be error checked. Nothing should have been changed
     * @param input
     * @return
     */
int verifyCRC (uint8_t input[], int sizeOfArray) {
	uint8_t actualMessage [HCE_MAX_MESSAGE];
	//get the actual message
	for (int i = 0; i < sizeOfArray; i++)
		actualMessage[i]=input[i+2];

	uint8_t crcBytes[2] = {input[0], input[1]};
	int crcFromMessage = combineBytes(crcBytes);
	int generated = generateUniTapCrc(actualMessage, sizeOfArray-2);
	return (crcFromMessage == generated);
}

    /**
     * Combine two bytes into one integer
     * @param bytes
     * @return
     */
int combineBytes(uint8_t bytes[]){
	return (int)( (bytes[1] << 8)|(bytes[0])&0xff );
}

    /**
     * Take a message and add a crc of the message to the beginning
     * @param input
     * @return
     */
uint8_t * appendCRCBytes(uint8_t input[], int sizeOfArray){
	static uint8_t output [HCE_MAX_MESSAGE] = {0x00};

	//generate crc of crc-target string
	int crcInt = generateUniTapCrc(input, sizeOfArray);

	//mask out the two important bytes
	
	output[0] = crcInt & 0xFF;
	output[1] = (crcInt >> 8) & 0xFF;

	//now merge those bytes with the original message
		//copy previous message
	for (int i = 0; i<sizeOfArray;i++){
		output[i+2]=input[i];
	}
	//return the concatenated message
	return output;
}

    /**
     * This will take one message (in byte array 1), and append a CRC generated from another (int byte array 2)
     * @param arrayToAppend the byte array to add the CRC to
     * @param arrayToCRC the byte array to generate the CRC from
     * @return the appended array
     */
uint8_t * appendCRCOnDifferentArray(uint8_t appendArray[], int appendArrayLength, uint8_t crcArray[], int crcArrayLength){
	static uint8_t output [HCE_MAX_MESSAGE]= {0X00};
	
	//generate crc of crc-target string
	int crcInt = generateUniTapCrc(crcArray, crcArrayLength);

	//mask out the two important bytes
	uint8_t crcBytes [2] = {};
	crcBytes[1] = (crcInt >> 8) & 0xFF;
	crcBytes[0] = crcInt & 0xFF;

	//now merge those bytes with the original message
	static int outputSize = 0;
	outputSize = appendArrayLength+2;
	
		//copy previous message
	for (int i = 0; i<outputSize;i++){
		output[i+2]=appendArray[0];
	}
		//add crc to message
	output[0] = crcBytes[0];
	output[1] = crcBytes[1];

	//return the concatenated message
	return output;
}